angular/lib/src/runtime/view.dart

```diff
@@ -0,0 +1,176 @@
+/// Proposal for generated application views for components and templates.
+///
+/// Not currently used in production.
+///
+/// See: https://github.com/dart-lang/angular/issues/1421
+library angular.src.runtime.view;
+
+import 'dart:html' show HtmlElement;
+
+import 'package:meta/meta.dart' show mustCallSuper, protected;
+
+import '../core/linker/view_container.dart';
+import '../di/injector/injector.dart';
+import 'optimizations.dart';
+
+/// Base generated view for components and templates.
+abstract class View<T> {
+  /// Runs change detection, updating the view and children based on context.
+  @mustCallSuper
+  void detectChanges() {}
+
+  /// Destroys this view and disposes of any resources or child views.
+  @mustCallSuper
+  void destroy() {
+
+  }
+}
+
+/// Base generated view for views that have a parent [ComponentView].
+abstract class ChildView<T> extends View<T> {
+  /// Concrete component `class` instance [T].
+  final T context;
+
+  ChildView(this.context);
+}
+
+/// Root application view for a given component class [T].
+///
+/// For example given:
+/// ```dart
+/// @Component(...)
+/// class A {}
+/// ```
+///
+/// ... would generate:
+///
+/// ```dart
+/// class _ComponentView$A extends ComponentView<A> {}
+/// ```
+abstract class ComponentView<T> extends ChildView<T> {
+  /// Child views (embedded or components).
+  @protected
+  final List<View<void>> childViews;
+
+  /// Root element used as the root element for the component view.
+  @protected
+  final HtmlElement rootElement;
+
+  ComponentView(
+    T context,
+    this.childViews,
+    this.rootElement,
+  ) : super(context);
+
+  @mustCallSuper
+  @override
+  void destroy() {
+    destroyViewChildren();
+    super.destroy();
+  }
+
+  @mustCallSuper
+  @protected
+  void destroyViewChildren() {
+    for (final childView in childViews) {
+      childView.destroy();
+    }
+  }
+}
+
+/// Nested embedded view for `<template>` views inside of a component [T].
+///
+/// Any time a `<template>` tag is inserted into the template of a _component_
+/// (explicitly, or implicitly via `*`-directives, `<ng-container>`, or
+/// annotations like `@deferred`), an [EmbeddedView] is created.
+///
+/// For example given:
+/// ```dart
+/// @Component(
+///   template: r'''
+///     <template [ngIf]="condition">...</template>
+///   ''',
+/// )
+/// class A {}
+/// ```
+///
+/// ... would generate:
+///
+/// ```dart
+/// class _EmbeddedView_0_NgIf$A extends EmbeddedView<A> {}
+/// ```
+abstract class EmbeddedView<T> extends ChildView<T> {
+  @protected
+  final ComponentView<T> componentView;
+
+  final Map<String, dynamic> _locals = Map.identity();
```

Comment:
Could we possible type `_locals` with a type parameter?

```dart
abstract class EmbeddedView<T, L> extends ChildView<T> {
  final L locals;
  ...
}
```

Then invocations of `embeddedView.getLocal(String name)` could be replaced with `embeddedView.locals.name`.

It's not clear to me whether or how this works in every context we need to invoke locals. For example, anything that accesses a local from the template would probably be covered, however how would client code get the `L` type? Would they inject a `Template<L>` or `ViewContainer<L>`?